package/MDAnalysis/lib/nsgrid.pyx

```diff
@@ -0,0 +1,588 @@
+        PyMem_Free(self.beadids)
+        PyMem_Free(self.cellids)
+
+    cdef ns_int coord2cellid(self, rvec coord) nogil:
+        return <ns_int> (coord[ZZ] / self.cellsize[ZZ]) * (self.cell_offsets[ZZ]) +\
+               <ns_int> (coord[YY] / self.cellsize[YY]) * self.cell_offsets[YY] + \
+               <ns_int> (coord[XX] / self.cellsize[XX])
+
+    cdef bint cellid2cellxyz(self, ns_int cellid, ivec cellxyz) nogil:
+        if cellid < 0:
+            return False
+        if cellid >= self.size:
+            return False
+
+        cellxyz[ZZ] = <ns_int> (cellid / self.cell_offsets[ZZ])
+        cellid -= cellxyz[ZZ] * self.cell_offsets[ZZ]
+
+        cellxyz[YY] = <ns_int> (cellid / self.cell_offsets[YY])
+        cellxyz[XX] = cellid - cellxyz[YY] * self.cell_offsets[YY]
+
+        return True
+
+    cdef fill_grid(self, real[:, ::1] coords):
+        cdef ns_int i, cellindex = -1
+        cdef ns_int ncoords = coords.shape[0]
+        cdef ns_int[:] beadcounts = np.empty(self.size, dtype=np.int)
+
+        with nogil:
+            # Initialize buffers
+            for i in range(self.size):
+                beadcounts[i] = 0
+
+            # First loop: find cellindex for each bead
+            for i in range(ncoords):
+                cellindex = self.coord2cellid(&coords[i, 0])
+
+                self.nbeads[cellindex] += 1
+                self.cellids[i] = cellindex
+
+                if self.nbeads[cellindex] > self.nbeads_per_cell:
+                    self.nbeads_per_cell = self.nbeads[cellindex]
+
+        # Allocate memory
+        self.beadids = <ns_int *> PyMem_Malloc(sizeof(ns_int) * self.size * self.nbeads_per_cell) #np.empty((self.size, nbeads_max), dtype=np.int)
+        if not self.beadids:
+            raise MemoryError("Could not allocate memory for NSGrid.beadids ({} bits requested)".format(sizeof(ns_int) * self.size * self.nbeads_per_cell))
+
+        with nogil:
+            # Second loop: fill grid
+            for i in range(ncoords):
+
+                # Add bead to grid cell
+                cellindex = self.cellids[i]
+                self.beadids[cellindex * self.nbeads_per_cell + beadcounts[cellindex]] = i
+                beadcounts[cellindex] += 1
+
+
+
+cdef class FastNS(object):
+   
+    cdef PBCBox box
+    cdef real[:, ::1] coords
+    cdef real[:, ::1] coords_bbox
+    cdef readonly real cutoff
+    cdef NSGrid grid
+    cdef ns_int max_gridsize
+    cdef bint periodic
+    
+
+    def __init__(self, cutoff, coords, box=None, max_gridsize=5000):
+        import MDAnalysis as mda
+        from MDAnalysis.lib.mdamath import triclinic_vectors
+
+        cdef real[:] pseudobox = np.zeros(6, dtype=np.float32)
+        cdef real[DIM] bmax, bmin
+        cdef ns_int i
+        self.periodic = True
+
+
+        if (box is None) or (np.allclose(box[:3], 0.) and box.shape[0] == 6):
+            bmax = np.max(coords, axis=0)
+            bmin = np.min(coords, axis=0)
+            for i in range(DIM):

```

Comment:
Mostly looping over all the coordinates to shift the origin since box origin is always (0,0,0) which is used to get the cellids later on. Similarly, np.min and np.max are also dark yellow if you look into annotate.